samples/gamma-fade/main.c

```diff
@@ -0,0 +1,268 @@
+        entries[i].green = 0xFF * f;
+        entries[i].blue  = 0xFF * f;
+    }
+    XVideoSetGammaRamp(0, entries, 256);
+}
+
+static bool animate_frame(DWORD ts_start, DWORD ts, float duration, float* _t)
+{
+    /* Wait for the next frame */
+    XVideoWaitForVBlank();
+
+    /* Update timestamp */
+    DWORD ts_delta = ts - ts_start;
+    float t = remap(0.0f, duration, 0.0f, 1.0f, ts_delta);
+
+    /* Signal completion */
+    if (t >= 1.0f) {
+      return false;
+    }
+
+    /* Writeback time */
+    *_t = t;
+    return true;
+}
+
+static void do_mix_fade(float red, float green, float blue,
+                        float(*interpolator)(float), float duration)
+{
+    DWORD ts_start;
+    float t;
+
+    /* Fade to color */
+    ts_start = GetTickCount();
+    while(animate_frame(ts_start, GetTickCount(), duration / 2.0f, &t)) {
+        mix_fade(red, green, blue,
+                 interpolator(t));
+    }
+
+    /* Fade to image */
+    ts_start = GetTickCount();
+    while(animate_frame(ts_start, GetTickCount(), duration / 2.0f, &t)) {
+        mix_fade(red, green, blue,
+                 interpolator(1.0f - t));
+    }
+
+    /* Reset gamma to avoid partial fade */
+    reset_gamma();
+}
+
+static void do_cinematic_fade(float a_black_input,  float a_white_input,
+                              float a_black_output, float a_white_output,
+                              float b_black_input,  float b_white_input,
+                              float b_black_output, float b_white_output,
+                              float(*interpolator)(float), float duration)
+{
+    DWORD ts_start;
+    float t;
+
+    /* Fade to color */
+    ts_start = GetTickCount();
+    while(animate_frame(ts_start, GetTickCount(), duration / 2.0f, &t)) {
+        cinematic_fade(a_black_input,  a_white_input,
+                       a_black_output, a_white_output,
+                       b_black_input,  b_white_input,
+                       b_black_output, b_white_output,
+                       interpolator(t));
+    }
+
+    /* Fade to image */
+    ts_start = GetTickCount();
+    while(animate_frame(ts_start, GetTickCount(), duration / 2.0f, &t)) {
+        cinematic_fade(a_black_input,  a_white_input,
+                       a_black_output, a_white_output,
+                       b_black_input,  b_white_input,
+                       b_black_output, b_white_output,
+                       interpolator(1.0f - t));
+    }
+
+    /* Reset gamma to avoid partial fade */
+    reset_gamma();
+}
+
+static void draw_title(const char* title)
+{
+    extern int nextCol;
+    extern int nextRow;
+    nextCol = 25;
+    nextRow = 25;
+    debugPrint("%s", title);
+    for(int i = strlen(title); i < 60; i++) {
+        debugPrint(" ");
+    }
+}
+
+int main(void)
+{
+    float duration = 2000.0f;
+
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    /* Create test image in framebuffer */
+    uint32_t *rgbx = (uint32_t*)XVideoGetFB();
```

Comment:
Probably want to make this `volatile`

Reply:
Why? This is also how it's done in the other gamma sample. The framebuffer memory is just standard RAM and the compiler shouldn't be able to skip memory writes (at least not until I hammer the same address with the same value).

Reply:
> Why? This is also how it's done in the other gamma sample.

I should have mentioned it when reviewing the other sample.

> The framebuffer memory is just standard RAM and the compiler shouldn't be able to skip memory writes (at least not until I hammer the same address with the same value)

Compilers can (and do) skip memory writes. Although we don't currently suffer this problem, in the future a sufficiently aggressive compiler optimization pass may chose to eliminate what it believes are dead writes because it lacks the knowledge that this memory is being sampled by the video encoder.

As an example, I'll turn on link-time optimization:
```c
int x[10];

int *get_fb()
{
    return x;
}

void draw()
{
    int *y = get_fb(); // a
    y[0] = 10;
    y[1] = 100;
}

int main(int argc, char *argv[])
{
    draw();
    return 0x1337;
}
```
``` bash
$ gcc -o example -O3 -flto example.c
$ objdump -dMintel example | grep -A4 main\>:
00000000000004f0 <main>:
 4f0:	b8 37 13 00 00       	mov    eax,0x1337
 4f5:	c3                   	ret    
 4f6:	66 2e 0f 1f 84 00 00 	nop    WORD PTR cs:[rax+rax*1+0x0]
 4fd:	00 00 00 
```

Changing `(a)` to be `volatile int *y = get_fb();`, we get instead what we expected:

```bash
$ objdump -dMintel example | grep -A6 main\>:
00000000000004f0 <main>:
 4f0:	c7 05 46 0b 20 00 0a 	mov    DWORD PTR [rip+0x200b46],0xa        # 201040 <x>
 4f7:	00 00 00 
 4fa:	b8 37 13 00 00       	mov    eax,0x1337
 4ff:	c7 05 3b 0b 20 00 64 	mov    DWORD PTR [rip+0x200b3b],0x64        # 201044 <x+0x4>
 506:	00 00 00 
 509:	c3                   	ret
```

Reply:
This isn't actually what's happening in this code though. The optimization is only allowed because the compiler can check if `x[0..9]` is accessed elsewhere (it isn't).

In our use-case, the compiler doesn't know what memory we are accessing (it's not managed by the compiler), so it can't check if anything uses it, and it won't make any assumptions.

So, for simulating what we do here, replace `int x[10];` by `int* x = (int*)0x100;`:

```
$ objdump -dMintel example | grep -A6 main\>
0000000000001020 <main>:
    1020:	48 b8 0a 00 00 00 64 	movabs rax,0x640000000a
    1027:	00 00 00 
    102a:	48 89 04 25 00 01 00 	mov    QWORD PTR ds:0x100,rax
    1031:	00 
    1032:	b8 37 13 00 00       	mov    eax,0x1337
    1037:	c3                   	ret    
```

*(Interestingly `int m[10]; int* x = m;` is also enough to break this optimization - the write still happens)*

I'd have to do more reading, but I'm confident that this should not require `volatile`. If it did, it would break thousands of applications which map device memory into the host. [D3D](https://docs.microsoft.com/en-us/windows/win32/direct3d9/d3dlocked-rect), [SDL2](https://hg.libsdl.org/SDL/file/default/include/SDL_surface.h#l76) or [GL](https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glMapBuffer.xml) don't use `volatile` in these situations either.

I do see the validity of your comment though. It *does* get interesting when adding a loop around those writes.

- `int x = 5; while(x--)` will only lead to a single write (which is somewhat expected and wanted), however, when adding some call or other action into the loop, it will do the writes again (I tested with `usleep(100)`).
- `while(1)` will lead to what I believe is a bug/oversight in gcc. Because gcc will then skip the writes - which breaks the assumption I (and also those other APIs) just did. However, clang actually keeps a single write before entering the endless loop.

> We should enforce this by also updating XVideoGetFB in a follow-up PR to return a volatile pointer.

I think we should be careful with `volatile` pointers in APIs, and expect the user to make logical decisions when accessing memory. `volatile` would still cause a lot of unintended sideffects (performance + incomplete cache flush), because the framebuffer will probably be write-combined in the future.

At the moment the discussion is also severely affected by an orthogonal issue, because the buffer is not even owned (#169). That is much worse than anything a lack of `volatile` could cause, and it will also require refactors in application code probably.

Most users will (or should) probably just use pbkit or SDL (or the ugly hal/debug.c) to access the framebuffer anyway.

---

For this PR, I feel like we should keep it consistent with the other gamma sample.

We should potentially have a follow-up to add the `volatile` or to replace such accesses by SDL (which can be assumed to be well-defined API which does not require `volatile`). Using SDL would also allow support for 16-bit modes out-of-the-box. I just wanted to avoid the additional dependency.

The pattern shown by the gamma samples should also be improved.

Reply:
> This isn't actually what's happening in this code though. [...]

I think this is a fair analog to what could happen in the future, even if it's not happening now with a pointer. In the past I've had compilers silently eliminate MMIO writes. While it may work now, I wouldn't be surprised to see this break in the future if (when) we apply whole-program optimization.

> If it did, it would break thousands of applications which map device memory into the host. D3D, SDL2 or GL don't use volatile in these situations either.

This is an interesting point, and I'd like to review the C memory rules more carefully--I believe the reason it works in the cases you've mentioned is due to these calls being made into shared libraries, the compiler obviously cannot optimize these writes away because this code is not loaded until runtime.

> I think we should be careful with volatile pointers in APIs, and expect the user to make logical decisions when accessing memory. volatile would still cause a lot of unintended sideffects (performance + incomplete cache flush), because the framebuffer will probably be write-combined in the future (or otherwise cached; volatile does not take care of flushes).

I agree, we *should* be careful. I'm not sure what specific side effects you're referring to in this case. The compiler is not responsible for flushing the cache--the processor itself (or OS) will handle that. The memory allocated for the frame buffer *should* be configured as write-combining, which will combine writes made by software, but the software itself (at least *this* software, OS memory management/context switching is different) doesn't handle flushing the cache.

> At the moment the discussion is also severely affected by an orthogonal issue, because the buffer is not even owned (#169). That is much worse than anything that a lack of volatile could cause, and it will also require refactors in application code probably.

I don't think this will require significant refactors to be honest. But you are correct, we should allocate the memory during `XVideoInit` (with appropriate flags) and maintain a reference to it.

As we are talking about changing to use allocated memory, here's another sample of dead write elimination that I found interesting:

```c
#include <stdlib.h>
#include <string.h>

int *get_fb()
{
    return malloc(8);
}

void draw()
{
    int *fb = get_fb();
    memset(fb, 0, 8);
}

int main(int argc, char **argv[])
{
    draw();
    return 0x1337;
}
```
```
00000000000004f0 <main>:
 4f0:	b8 37 13 00 00       	mov    eax,0x1337
 4f5:	c3                   	ret    
```



> Most users will (or should) probably just use pbkit or SDL (or the ugly hal/debug.c) to access the framebuffer anyway.

Possibly? Possibly not. This is the API we have today, and we should improve it or get rid of it.

> For this PR, I feel like we should keep it consistent with the other gamma sample.

I disagree. I think we should fix it here now, and go back and fix it in the other sample as well. This sample may work today with the compiler and settings we are using, but I don't expect it to work in the future. It's a simple change to guarantee that the compiler preserves these writes to the framebuffer and we should do it.

Reply:
> This isn't actually what's happening in this code though. [...]

I think this is a fair analog to what could happen in the future, even if it's not happening now with a pointer. In the past I've had compilers silently eliminate MMIO writes (this will typically happen if you write to a location twice, the compiler just writes once with the second value--but if I recall correctly, I've also had it outright just drop a single write). While it may work now, I wouldn't be surprised to see this break in the future if (when) we apply whole-program optimization.

> If it did, it would break thousands of applications which map device memory into the host. D3D, SDL2 or GL don't use volatile in these situations either.

This is an interesting point, and I'd like to review the C memory rules more carefully--I believe the reason it works in the cases you've mentioned is due to these calls being made into shared libraries, the compiler obviously cannot optimize these writes away because this code is not loaded until runtime. I also think if memory is mapped into the host address space, code should be very cautious about compilers optimizing away writes to shared buffers, which may also include use of the `volatile` keyword. Frankly, I need to review the C spec to better understand the semantics of these dead-write elimination optimizations.

> I think we should be careful with volatile pointers in APIs, and expect the user to make logical decisions when accessing memory. volatile would still cause a lot of unintended sideffects (performance + incomplete cache flush), because the framebuffer will probably be write-combined in the future (or otherwise cached; volatile does not take care of flushes).

I agree, we *should* be careful. I'm not sure what specific side effects you're referring to in this case. The compiler is not responsible for flushing the cache--the processor itself (or OS) will handle that. The caching policy for the memory allocated for the frame buffer *should* be configured as write-combining, which will combine writes made by software, but the software itself (at least *this* software, OS memory management/context switching is different) doesn't typically handle flushing the cache, although there certainly may be some times when we probably want our "driver level" software to do that.

> At the moment the discussion is also severely affected by an orthogonal issue, because the buffer is not even owned (#169). That is much worse than anything that a lack of volatile could cause, and it will also require refactors in application code probably.

I don't think this will require significant refactors to be honest. But you are correct, we should allocate the memory during `XVideoInit` (with appropriate flags)--or something along those lines--and maintain a reference to it.

As we are talking about changing to use allocated memory, here's another sample of dead write elimination that I found interesting:

```c
#include <stdlib.h>
#include <string.h>

int *get_fb()
{
    return malloc(8);
}

void draw()
{
    int *fb = get_fb();
    memset(fb, 0, 8);
}

int main(int argc, char *argv[])
{
    draw();
    return 0x1337;
}
```
```
00000000000004f0 <main>:
 4f0:	b8 37 13 00 00       	mov    eax,0x1337
 4f5:	c3                   	ret    
```

Here you can see I have switched it over to using a pointer, and the compiler realized that these writes are not useful and rightly eliminated the allocation and write. You can add sufficient complexity around here to get the compiler to not do this, but eventually as compilers get even better, I think such elimination will still be done. Perhaps surprisingly in this case, *even using `volatile`* doesn't fix this elimination (the compiler does rightfully complain when passing this pointer to `memset` discards the `volatile` keyword; it ignores the `volatile` keyword in this case and respects the non-volatile definition and then proceeds to eliminate the writes).

If I change the code to not use `memset`, but just dereference and assign to the memory, then `volatile` works as expected, and the `volatile` semantics are preserved:

```c
#include <stdlib.h>
#include <string.h>

int *get_fb()
{
    return malloc(8);
}

void draw()
{
    int *fb = get_fb();
    *fb = 20;
}

int main(int argc, char *argv[])
{
    draw();
    return 0x1337;
}
```

Without `volatile`:
```
0000000000000540 <main>:
 540:	b8 37 13 00 00       	mov    eax,0x1337
 545:	c3                   	ret    
```

and with:
```
0000000000000540 <main>:
 540:	48 83 ec 08          	sub    rsp,0x8
 544:	bf 08 00 00 00       	mov    edi,0x8
 549:	e8 d2 ff ff ff       	call   520 <malloc@plt>
 54e:	c7 00 14 00 00 00    	mov    DWORD PTR [rax],0x14
 554:	b8 37 13 00 00       	mov    eax,0x1337
 559:	48 83 c4 08          	add    rsp,0x8
 55d:	c3                   	ret
```

> Most users will (or should) probably just use pbkit or SDL (or the ugly hal/debug.c) to access the framebuffer anyway.

Possibly? Possibly not. This is the API we have today, and we should improve it or get rid of it.

> For this PR, I feel like we should keep it consistent with the other gamma sample.

I disagree. I think we should fix it here now, and go back and fix it in the other sample as well. This sample may work today with the compiler and settings we are using, but I don't expect it to work in the future. It's a simple change to guarantee that the compiler preserves these writes to the framebuffer and we *should* do it.

Reply:
I added a new commit to make the gamma sample accessor `volatile` and also made it `volatile` here.

Reply:
Thanks